pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,236 @@
+		}
+	case !existInAtlas && wasDeleted:
+		return r.unmanage(ctx, akoPrivateEndpoint)
+	case existInAtlas && wasDeleted:
+		if customresource.IsResourcePolicyKeepOrDefault(akoPrivateEndpoint, r.ObjectDeletionProtection) {
+			return r.unmanage(ctx, akoPrivateEndpoint)
+		}
+
+		atlasPEService, err = r.deletePrivateEndpoint(ctx.Context, projectID, atlasPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointFailedToDelete, err)
+		}
+	}
+
+	switch atlasPEService.Status() {
+	case privateendpoint.StatusInitiating:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is being initialized")
+	case privateendpoint.StatusPending, privateendpoint.StatusPendingAcceptance, privateendpoint.StatusWaitingForUser, privateendpoint.StatusVerified:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is waiting for human action")
+	case privateendpoint.StatusFailed, privateendpoint.StatusRejected:
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToConfigure, errors.New(atlasPEService.ErrorMessage()))
+	case privateendpoint.StatusDeleting:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceDeleting, "Private Endpoint is being deleted")
+	}
+
+	ctx.SetConditionTrue(api.PrivateEndpointServiceReady)
+	r.EventRecorder.Event(akoPrivateEndpoint, "Normal", string(workflow.PrivateEndpointServiceCreated), "Private Endpoint Service is available")
+
+	return r.handlePrivateEndpointInterface(ctx, projectID, akoPrivateEndpoint, akoPEService, atlasPEService)
+}
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointInterface(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+	akoPEService privateendpoint.EndpointService,
+	atlasPEService privateendpoint.EndpointService,
+) (ctrl.Result, error) {
+	compositeInterfacesMap := privateendpoint.MapPrivateEndpoints(akoPEService.EndpointInterfaces(), atlasPEService.EndpointInterfaces())
+
+	if len(compositeInterfacesMap) == 0 {
+		return r.waitForConfiguration(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	// we want to sync all interface, if any of them is in progress, after all we transition to in progress
+	pendingResources := false
+	for _, compositeInterfaceMap := range compositeInterfacesMap {
+		inAKO := compositeInterfaceMap.AKO != nil
+		inAtlas := compositeInterfaceMap.Atlas != nil
+		inProgress := isInterfaceInProgress(compositeInterfaceMap.Atlas)
+		failed := hasInterfaceFailed(compositeInterfaceMap.Atlas)
```

Comment:
We have 4 flags, which could render 16 combinations, but only 6 transitions in the state machine. It would be good to point here to the state machine diagram or summarise it.

Reply:
The number of flags is not directly proportional to the number of transitions, once some applies only on a specific state.
Anyway, I'm gonna to add a doc block here explaining it